madSimonJ/CSharpHacks
# Various Additional Extension Methods

- **CSharpHacks Project:**
     - **Update:** Target Framework updated to `netstandard2.0`. 
     - **Update:** C# Language Version updated to 10.0.
     - **Added:** InternalsVisibleTo attribute to allow CSharpHacks.Tests access to internal scope types and members.

 - **CSharpHacks.Tests Project:**
    - **Update:** Target Framework updated to `net7.0`. Support for `INumber<T>` for numerical test cases.
    - **Update:** C# Language Version updated to 10.0. 
    - **Update:** Updated Nuget package reference versions.
    - **Added:** Unit Tests for all added extension methods, below. 
    - **Added:** Mock objects for tests that need them. 
    - **Refactor:** Moved helper classes to sub-folder.

 - **ObjectHacks:** 
     - **Added:** `DynamicProperties.` Gets a dynamic collection of properties associated with an object instance, with a lifetime scoped to the lifetime of the object. 
     - **Added:** `To.` Directly casts the object instance to a specified type. 
     - **Added:** `As.` Safely casts the object instance to a specified type.

 - **TaskHacks:**
    - **Added:** `OnCompletedSuccessfully.` Executes an action when a task has been completed successfully.

 - **AssemblyExtensions:** 
     - **Added:** `GetAllTypesImplementingOpenGenericType.` Gets a collection of all types implementing a specified open generic interface, within a given assembly. 
     - **Added:** `InstantiateAllTypes.` Scans an assembly for all instantiable classes of a specified base class, and forms an array of instances. Note: Doesn't work with interfaces.

 - **NumericalExtensions:**
    - **Added:** `IsWithinRange.` Determines whether a value with within a given inclusive range.
    - **Added:** `IsGreaterThan.` Determines whether a value is greater than a given minimum value. 
    - **Added:** `IsLessThan.` Determines whether a value is less than a given maximum value. 
    - **Added:** `IsGreaterThanOrEqualTo.` Determines whether a value is greater than, or equal to, a given minimum value. 
    - **Added:** `IsLessThanOrEqualTo.` Determines whether a value is less than, or equal to, a given maximum value. 
    - **Added:** `FormatLargeNumber.` Formats the number in text form, with numerical group separators, for a given culture locale.

 - **EnumExtensions:** 
     - **Added:** `GetDescription.` Gets the description for the enum member, decorated with a `DescriptionAttribute`.

 - **EnumEx:**
    - **Added:** `Count.` Gets the number of values within this enum.

 - **EnumerableExtensions:** 
     - **Added:** `Random.` Chooses a value, at random, from a list of values. 
     - **Added:** `AddOrUpdate.` Adds a range of entries into a sorted dictionary. Includes a work around the fact that keys within Sorted Dictionaries cannot normally be overwritten. 
     - **Added:** `AddOrUpdateRange.` Adds a range of entries into a normal, or sorted dictionary. Includes a work around the fact that keys within Sorted Dictionaries cannot normally be overwritten. 
     - **Added:** `EnsureExistence.` If the value should exist, and it doesn't, it will be removed from the collection. If the value shouldn't exist, and it does, it will be removed from the collection. 
     - **Added:** `AddIfNotPresent.` Adds an item, if it not already present in the collection. 
     - **Added:** `RemoveIfPresent.` Removes an item, if it is already present in the collection. 
     - **Added:** `ContainsAny.` Determines whether the given collection contains any value within a given list of values. 
     - **Added:** `ContainsAnyKey.` Determines whether the given dictionary contains any key within a given list of keys. 
     - **Added:** `ContainsAnyValue.` Determines whether the given dictionary contains any value within a given list of values.

 - **ColourExtensions:**
    - **Added:** `UpdateColourChannel.` Updates a single colour channel within a `System.Drawing.Color` instance.
    - **Added:** `ToNormalisedRgba.` Normalises the specified colour. That being, it converts each channel value into a `double`, between 0.0 and 1.0, by dividing each value by `255d`. 
    - **Added:** `ToNormalisedArgb.` Normalises the specified colour. 
    - **Added:** `ToRgbaDoubleArray.` Converts a `Color` to an array of doubles, in RGBA order. 
    - **Added:** `ToArgbDoubleArray.` Converts a `Color` to an array of doubles, in ARGB order. 
    - **Added:** `ToRgbHexString.` Converts a `Color` to an RGB Hexadecimal string. 
    - **Added:** `ToArgbHexString.` Converts a `Color` to an ARGB Hexadecimal string. 
    - **Added:** `ToRgbaHexString.` Converts a `Color` to an RGBA Hexadecimal string. 
    - **Added:** `ToRgbString.` Converts a `Color` to an RGB() string.

 - **StringHacks:**
    - **Fixed:** Spelling mistakes within parameters, in various methods.
    - **Added:** `IfNullOrEmpty.` Returns a default string, if a specified string is null, or empty.
    - **Added:** `IfNullOrWhitespace.` Returns a default string, if a specified string is null, empty, or consists only of white-space characters.
    - **Added:** `ToOrdinalNumeral.` Converts a number to its ordinal numeral string representation. NOTE: Not localised.
    - **Added:** `ContainsLetters.` Determines whether a string contains any uppercase, or lowercase letters. Supports Unicode.
    - **Added:** `OnlyContainsLetters.` Determines whether a string contains any uppercase, or lowercase letters. Supports Unicode.
    - **Added:** `ContainsNumbers.` Determines whether a string contains any numbers. Supports Unicode.
    - **Added:** `ContainsOnlyNumbers.` Determines whether a string only contains any numbers. Supports Unicode.
    - **Added:** `GetDigits.` Strips out non-numeric characters in string, returning only digits. Supports Unicode.
    - **Added:** `GetLetters.` Strips out numeric and special characters in string, returning only letters. Supports Unicode.
    - **Added:** `GetAlphaNumerics.` Strips out any special characters in string, returning only letters and numbers. Supports Unicode.
    - **Added:** `GetNonAlphaNumerics.` Strips out any letters and numbers in string, returning only non-alpha-numeric characters. Supports Unicode.
    - **Added:** `SplitPascalCase.` Splits a string into sentence form, from a pascal case word.
    - **Added:** `EnsureEndsWith.` Appends a suffix to a string, if the suffix does not already appear at the end of the string.
    - **Added:** `EnsureStartsWith.` Prepends a prefix to a string, if the prefix does not already appear at the start of the string.
    - **Added:** `Pluralise.` Pluralises the specified string, based on a input value.
    - **Added:** `StartsWithAnyOf.` Determines whether a string starts with any strings within a given set.
    - **Added:** `StartsWithNoneOf.` Determines whether a string starts with any strings within a given set.
    - **Added:** `ContainsAnyOf.` Determines whether a string contains any strings within a given set.
    - **Added:** `ContainsNoneOf.` Determines whether a string contains any strings within a given set.
    - **Added:** `EndsWithAnyOf.` Determines whether a string ends with any strings within a given set.
    - **Added:** `EndsWithNoneOf.` Determines whether a string ends with any strings within a given set.
